src/cargo/ops/resolve.rs

```diff
@@ -49,6 +49,14 @@ pub fn resolve_ws_precisely<'a>(ws: &Workspace<'a>,
 
         add_overrides(&mut registry, ws)?;
 
+        for &(ref replace_spec, ref dep) in ws.root_replace() {
+            if !resolve.iter().any(|r| replace_spec.matches(r) && !dep.matches_id(r)) {
```

Comment:
Not sure that this is 100% correct: the idea is that replacement should match at least one package in the graph, but we also want to filter out the cases where replacement matches only itself. 

So for something like

```
[replace]
foo = bar
bar = foo
```

this won't report an error, but presumably such cases are unrealistic.